plugins/targetlocker/dblocker/dblocker.go

```diff
@@ -0,0 +1,310 @@
+func targetIDList(targets []*target.Target) []string {
+	res := make([]string, 0, len(targets))
+	for _, target := range targets {
+		res = append(res, target.ID)
+	}
+	return res
+}
+
+// listQueryString is a helper to create a (?, ?, ?) string
+// with as many ? as requested
+func listQueryString(length int) string {
+	switch length {
+	case 0:
+		return "()"
+	case 1:
+		return "(?)"
+	default:
+		return "(" + strings.Repeat("?, ", length-1) + "?)"
+	}
+}
+
+// DBLocker implements a simple target locker based on a relational database.
+// The current implementation only supports MySQL officially.
+// All functions in DBLocker are safe for concurrent by multiple goroutines.
+type DBLocker struct {
+	driverName string
+	db         *sql.DB
+	// lockTimeout set on each initial lock request
+	lockTimeout time.Duration
+	// refreshTimeout is used during refresh
+	refreshTimeout time.Duration
+}
+
+// cleanExpired deletes all expired locks on the given targets
+// (lock owner is ignored)
+func (d *DBLocker) cleanExpired(tx *sql.Tx, targets []string, now time.Time) error {
+	q := "DELETE FROM locks WHERE expires_at < ? AND target_id IN " + listQueryString(len(targets)) + ";"
+	// convert targets to a list of interface{}
+	queryList := make([]interface{}, 0, len(targets)+1)
+	queryList = append(queryList, now)
+	for _, targetId := range targets {
+		queryList = append(queryList, targetId)
+	}
+	_, err := tx.Exec(q, queryList...)
+	if err != nil {
+		return fmt.Errorf("unable to clean existing locks: %w", err)
+	}
+	return nil
+}
+
+// queryLocks returns a map of ID -> dblock for a given list of targets
+func (d *DBLocker) queryLocks(tx *sql.Tx, targets []string) (map[string]dblock, error) {
+	q := "SELECT target_id, job_id, expires_at FROM locks WHERE target_id IN " + listQueryString(len(targets))
+	// convert targets to a list of interface{}
+	queryList := make([]interface{}, 0, len(targets))
+	for _, targetId := range targets {
+		queryList = append(queryList, targetId)
+	}
+
+	rows, err := tx.Query(q, queryList...)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read existing locks: %w", err)
+	}
+	defer rows.Close()
+
+	row := dblock{}
+	locks := make(map[string]dblock)
+	for rows.Next() {
+		if err := rows.Scan(&row.targetId, &row.jobID, &row.expiresAt); err != nil {
+			return nil, fmt.Errorf("unexpected read from database: %w", err)
+		}
+		locks[row.targetId] = row
+	}
+	return locks, nil
+}
+
+// handleLock does the real locking, it assumes the jobID is valid
+func (d *DBLocker) handleLock(jobID int64, targets []string, timeout time.Duration) error {
+	// everything operates on this frozen time
+	now := time.Now()
+	expires := now.Add(timeout)
+	// locking is all or nothing in one transaction
+	// phantom reads might mess with the expired lock cleaning assumptions below,
+	// so request serializable isolation
+	tx, err := d.db.BeginTx(context.Background(), &sql.TxOptions{Isolation: sql.LevelSerializable})
```

Comment:
I'm no SQL expert - does `sql.LevelSerializable` also guarantee the desired consistency level? I.e. that  lock and subsequent read will always be consistent on any primary/replica?
At a quick look isolation and consistency should not be confused, and I can't confirm that this is what we want for DB locking

Reply:
Not sure I am following here. I am using this purely for query isolation, since some steps assume previous steps are done properly. Example: When locking I read other locks and check them (because we need this to produce a useful error anyway). The DB will enforce only one lock per asset on its own of course, but it makes the code simple to assume there won't be any conflicts after I run the check.

What are you asking about exactly wrt. to consistency, do you mean strict read-after-write consistency for multiple queries to the same DB?

Reply:
we discussed this internally already, and concluded that transactions paired to serializable isolation should be good enough for locking purposes. To answer to your question, yes, I meant strict read-after-write consistency